cpp/src/parquet/column_reader_test.cc

```diff
@@ -703,9 +710,187 @@ class RecordReaderTest : public ::testing::Test {
+  if (GetParam() == /*read_dense_for_nullable=*/true) {
+    CheckState(/*values_written=*/0, /*null_count=*/0, /*levels_written=*/1,
+               /*levels_position=*/1);
+    CheckReadValues(/*expected_values=*/{},
+                    /*expected_defs=*/{0},
+                    /*expected_reps=*/{});
+  } else {
+    CheckState(/*values_written=*/1, /*null_count=*/1, /*levels_written=*/1,
+               /*levels_position=*/1);
+    CheckReadValues(/*expected_values=*/{kNullValue},
+                    /*expected_defs=*/{0},
+                    /*expected_reps=*/{});
+  }
+  record_reader_->Reset();
+  CheckState(/*values_written=*/0, /*null_count=*/0, /*levels_written=*/0,
+             /*levels_position=*/0);
+}
+
+// Tests reading a required repeated field. The results are the same for reading
+// dense or spaced.
+TEST_P(RecordReaderPrimitiveTypeTest, ReadRequiredRepeated) {
+  // Set repeated_ancestor_def_level = max_def_level so that the repeated field
+  // is not nullable.
+  Init(/*max_def_level=*/1, /*max_rep_level=*/1, Repetition::REPEATED,
+       /*repeated_ancestor_def_level=*/1);
+
+  // Records look like: {[10], [20, 20], [30, 30, 30]}
+  std::vector<std::shared_ptr<Page>> pages;
+  std::vector<int32_t> values = {10, 20, 20, 30, 30, 30};
+  std::vector<int16_t> def_levels = {1, 1, 1, 1, 1, 1};
+  std::vector<int16_t> rep_levels = {0, 0, 1, 0, 1, 1};
+
+  std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+      descr_.get(), values, /*num_values=*/static_cast<int>(def_levels.size()),
+      Encoding::PLAIN,
+      /*indices=*/{},
+      /*indices_size=*/0, def_levels, level_info_.def_level, rep_levels,
+      level_info_.rep_level);
+  pages.push_back(std::move(page));
+  auto pager = std::make_unique<MockPageReader>(pages);
+  record_reader_->SetPageReader(std::move(pager));
+
+  // Read [10]
+  int64_t records_read = record_reader_->ReadRecords(/*num_records=*/1);
+  ASSERT_EQ(records_read, 1);
+  CheckState(/*values_written=*/1, /*null_count=*/0, /*levels_written=*/6,
+             /*levels_position=*/1);
+  CheckReadValues(/*expected_values=*/{10}, /*expected_defs=*/{1},
+                  /*expected_reps=*/{0});
+  record_reader_->Reset();
+  CheckState(/*values_written=*/0, /*null_count=*/0, /*levels_written=*/5,
+             /*levels_position=*/0);
+
+  // Read [20, 20], [30, 30, 30]
+  records_read = record_reader_->ReadRecords(/*num_records=*/3);
+  ASSERT_EQ(records_read, 2);
+  CheckState(/*values_written=*/5, /*null_count=*/0, /*levels_written=*/5,
+             /*levels_position=*/5);
+  CheckReadValues(/*expected_values=*/{20, 20, 30, 30, 30},
+                  /*expected_defs=*/{1, 1, 1, 1, 1},
+                  /*expected_reps=*/{0, 1, 0, 1, 1});
+  record_reader_->Reset();
+  CheckState(/*values_written=*/0, /*null_count=*/0, /*levels_written=*/0,
+             /*levels_position=*/0);
+}
+
+// Tests reading a nullable repeated field.
+TEST_P(RecordReaderPrimitiveTypeTest, ReadNullableRepeated) {
```

Comment:
Could you please add a test case for `empty list` and `list of null`?
For example:
- {[10], [] , [20, 20], [], [30, 30, 30]}
- {[10], [null], [20, 20], [30, null, 30], [null, null]}

So we can make sure cases for `max_def_level > 1` are well-covered.

Reply:
I modified ReadNullableRepeated to cover these cases.